java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -499,6 +499,18 @@ class SummaryCall extends DataFlowCall, TSummaryCall {
   override Location getLocation() { result = c.getLocation() }
 }
 
+private predicate id(BasicBlock x, BasicBlock y) { x = y }
+
+private predicate idOf(BasicBlock x, int y) = equivalenceRelation(id/2)(x, y)
+
+class NodeRegion instanceof BasicBlock {
+  string toString() { result = "NodeRegion" }
```

Comment:
That feels slightly wrong to me, since a `NodeRegion` isn't necessarily a basic block - it could also be defined as e.g. a set of nodes guarded by a given guard. Currently the data flow side of the implementation assumes that `NodeRegion`s are disjoint, but that doesn't have to be the case. It could conceivably be useful to define them differently if we run into guards that guard a huge number of basic blocks.

Reply:
But its are purely internal implementation detail, right, so we can always change it as we see fit? I merely suggested to use `final extends` to avoid the dummy `toString` definitions.

Reply:
`final extends` needs a dummy alias instead, so there's not much saved. And it still feels wrong for the stated reason: A basic block is certainly a region of nodes, but a region of nodes is not necessarily a basic block.